File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.java/core/src/test/java/de/unistuttgart/hamster/UndoCommandTest.java
package de.unistuttgart.hamster;

import de.unistuttgart.hamster.hamster.HamsterGame;
import de.unistuttgart.hamster.util.GameStringifier;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class UndoCommandTest {
	
	@Test
	public void testUndo() {
		HamsterGame game = GameStringifier.createFromString(" >*;" +
															"   ;");
		game.hardReset();
		game.startGame();

		var hamster = game.getTerritory().getDefaultHamster();
		hamster.move();
		hamster.pickGrain();

		hamster.turnLeft();
		hamster.turnLeft();
		hamster.turnLeft();

		hamster.move();

		var commandStack = game.getGameCommandStack();
		commandStack.undo();
		String actual = GameStringifier.toString(game);
		assertEquals("  v;" +
				     "   ;", actual);

		commandStack.undoAll();

		actual = GameStringifier.toString(game);
		assertEquals(" >*;" +
				     "   ;", actual);
	}

}

File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/test/UndoCommandTest.cpp
#include "gtest/gtest.h"

#include "HamsterCore.h"

#include "util/GameStringifier.h"
#include "util/TestUtils.h"

#include <utility>
#include <vector>
#include <string>

using namespace mpw;
using namespace hamster;
using namespace framework;
using namespace util;

/// \note use 'NO-LINT' comment at tests to suppress a warning caused by TEST_F
class UndoCommandTest : public testing::Test {
public:
};

//<editor-fold desc="Feature: undo">

TEST_F(UndoCommandTest, testUndo123) {
    std::shared_ptr<HamsterGame> game = GameStringifier::createFromString(" >*;"
                                                                          "   ;");
    game->hardReset();
    game->startGame();

    auto hamster = game->getTerritory()->getDefaultHamster();
    hamster->move();

    std::string actual1 = GameStringifier::toString(*game);
    EXPECT_EQ("  >;"
              "   ;", actual1);

    auto commandStack = game->getGameCommandStack();
    commandStack->undo();

    std::string actual = GameStringifier::toString(*game);
    EXPECT_EQ(" >*;"
              "   ;", actual);
}

TEST_F(UndoCommandTest, testUndo) { /* NOLINT */
    std::shared_ptr<HamsterGame> game = GameStringifier::createFromString(" >*;"
                                                                          "   ;");
    game->hardReset();
    game->startGame();

    auto hamster = game->getTerritory()->getDefaultHamster();
    hamster->move();
    hamster->pickGrain();

    hamster->turnLeft();
    hamster->turnLeft();
    hamster->turnLeft();

    hamster->move();

    auto commandStack = game->getGameCommandStack();
    commandStack->undo();
    std::string actual = GameStringifier::toString(*game);
    EXPECT_EQ("  v;"
              "   ;", actual);

    commandStack->undoAll();

    actual = GameStringifier::toString(*game);
    EXPECT_EQ(" >*;"
              "   ;", actual);
}

//</editor-fold>
